File: leetcode/array/18.py
from typing import List


class Solution:

    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        '''a+b+c+d=target的不重复解
        
        @Note:
            排序,四指针,固定两个,剩余两个双边遍历
        '''
        ret = []
        lens = len(nums)
        if lens < 4:
            return []
        nums = sorted(nums)
        for i in range(lens):  #当前循环获取当前值下所有合适解
            # if nums[i]>target:#与target=0不同
            #     break
            if i > 0 and nums[i] == nums[i - 1]:  #注意边界
                continue
            for j in range(i + 1, lens):  #当前循环获取当前值下所有合适解
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l = j + 1
                r = lens - 1
                while l < r:
                    sum = nums[i] + nums[j] + nums[l] + nums[r]
                    if sum == target:
                        ret.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l + 1] == nums[l]:  #下一值与当前值相等(去重)
                            l += 1
                        l + 1
                        while l < r and nums[r - 1] == nums[r]:
                            r -= 1
                        r -= 1
                    elif sum > target:
                        r -= 1
                    else:
                        l += 1
        return ret


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    target = int(input().strip())
    solution = Solution()
    print(solution.fourSum(arr, target))

File: leetcode/array/18.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
 public:
  void four_sum(std::vector<int>& arr, int target, std::vector<std::vector<int>>* ret) {
    std::sort(arr.begin(), arr.end());
    for (int i = 0; i < arr.size(); ++i) {
      if (arr[i] == arr[i - 1] && i != 0) { // 去重
        continue;
      }
      for (int j = i + 1; j < arr.size(); ++j) {
        if (arr[j] == arr[j - 1] && j != i + 1) { // 去重
          continue;
        }
        int k = j + 1, l = arr.size() - 1;
        while (k < l) {
          int sum = arr[i] + arr[j] + arr[k] + arr[l];
          if (sum == target) {
            std::vector<int> tmp = {arr[i], arr[j], arr[k], arr[l]};
            ret->emplace_back(tmp);
            while (arr[k] == tmp[2]) { // 去重
              ++k;
            }
            while (arr[l] == tmp[3]) { // 去重
              --l;
            }
          } else if (sum < target) {
            ++k;
          } else {
            --l;
          }
        }
      }
    }
  }
};

int main() {
  std::vector<int> a = {1, 5, 0, 2, 2, 3};

  Solution solution = Solution();
  std::vector<std::vector<int>> ret;
  solution.four_sum(a, 5, &ret);
  for (const auto& items : ret) {
    for (const auto& item : items) {
      std::cout << item << " ";
    }
    std::cout << std::endl;
  }
}